Sensor - processing blocks - Refactoring

arilowen opened the pull request:
WIP:
- Fisheye auto-exposure fix

Commit:
rename foreach to foreachRS

dorodnic committed:
Fixing python wrapper

alowenst01 committed:
wip

alowenst01 committed:
wip options

alowenst01 committed:
wip - options 2

alowenst01 committed:
wip 3

alowenst01 committed:
wip - refactor ds5 devices with synthetic stream

alowenst01 committed:
wip - multiple requests

alowenst01 committed:
wip - dual y8 stream

alowenst01 committed:
wip - add left / right ir separate support

alowenst01 committed:
wip - fix source's sensor source

alowenst01 committed:
wip - fix multiple left - right IR start stop

alowenst01 committed:
wip - add mapping to init stream profiles

alowenst01 committed:
wip - sort profiles after init stream profiles

alowenst01 committed:
wip - best matching processing block

alowenst01 committed:
wip - choose best processing block

alowenst01 committed:
wip - composite processing block

alowenst01 committed:
wip - fix resolve many to many issues

alowenst01 committed:
wip - refactor resolve requests method

alowenst01 committed:
wip - add get frame data size to frame_interface

alowenst01 committed:
wip - add l500 support

alowenst01 committed:
fix merge conflicts

matkatz committed:
l500 fw update enabled. fw update serial number added.

alowenst01 committed:
wip - add l500 y8 rotation support

matkatz committed:
fw update id - cr fix

alowenst01 committed:
wip - l500 add zero order

alowenst01 committed:
wip - refactor start() output callback

alowenst01 committed:
wip - add l500 confidence

alowenst01 committed:
wip - correlate cloned profile to source profile

matkatz committed:
block l500 flash update

alowenst01 committed:
wip - refactor yuy2 converters

alowenst01 committed:
wip - refactor uvc open function

alowenst01 committed:
wip - hid refactoring

alowenst01 committed:
wip - hid sensor support

alowenst01 committed:
wip

fbagci committed:
Fixing typo in readme.md

alowenst01 committed:
wip - working hid

alowenst01 committed:
wip - multiple hid streams

alowenst01 committed:
wip - rename to motion-transform

alowenst01 committed:
wip - merge stream_info with stream_profile

alowenst01 committed:
wip - refactor color formats converter

alowenst01 committed:
wip - refactor l500-transform

alowenst01 committed:
wip - remove unnecessary files

alowenst01 committed:
wip - fix uid copy upon clone

aangerma committed:
Added to align sse support on RS2_DISTORTION_INVERSE_BROWN_CONRADY

alowenst01 committed:
wip - add bypass option to synthetic sensor

aangerma committed:
Fixed bug RS5-3373 , if the record stopped before recording camera info, viewer crashed because RS2_CAMERA_INFO_NAME doesn't supported by playback device.

alowenst01 committed:
wip - fix color options

ev-mp committed:
Merge pull request #4886 from fburakk/master

Fixing typo in readme.md

alowenst01 committed:
wip - fix motion option and remove on before frame callback

alowenst01 committed:
wip - rename sensors names

alowenst01 committed:
wip - cosmetics

alowenst01 committed:
wip - small fixes

alowenst01 committed:
merge with development

alowenst01 committed:
revert rs_capture example

alowenst01 committed:
remove confidence-to-raw8

aangerma commented on src/ds5/advanced_mode/advanced_mode.cpp:
```diff
@@ -239,9 +240,10 @@ namespace librealsense
+    bool ds5_advanced_mode_base::supports_option(const synthetic_sensor& sensor, rs2_option opt) const
     {
-        return sensor.supports_option(opt);
+        uvc_sensor* raw_uvc_sensor = dynamic_cast<uvc_sensor*>(sensor.get_raw_sensor().get());
```
why do you need to get the raw sensor? the synthetic_sensor support the options API

arilowen replied:
nice, thanks

aangerma commented on src/ds5/advanced_mode/advanced_mode.cpp:
```diff
@@ -540,14 +542,17 @@ namespace librealsense
+        uvc_sensor* raw_sensor = dynamic_cast<uvc_sensor*>((*_color_sensor)->get_raw_sensor().get());
         if (val.was_set)
-            _depth_sensor.get_option(RS2_OPTION_GAIN).set(val.gain);
+            raw_sensor->get_option(RS2_OPTION_GAIN).set(val.gain);
```
same as above

aangerma commented on src/ds5/advanced_mode/advanced_mode.cpp:
```diff
@@ -540,14 +542,17 @@ namespace librealsense
+
         if (val.was_set)
-            _depth_sensor.get_option(RS2_OPTION_ENABLE_AUTO_WHITE_BALANCE).set(float(val.auto_white_balance));
+            raw_sensor->get_option(RS2_OPTION_ENABLE_AUTO_WHITE_BALANCE).set(float(val.auto_white_balance));
```
same

ev-mp commented on include/librealsense2/h/rs_frame.h:
```diff
@@ -255,6 +255,23 @@ int rs2_is_frame_extendable_to(const rs2_frame* frame, rs2_extension extension_t
+* \param[in] source      Frame pool to allocate the frame from
+* \param[in] new_stream  New stream profile to assign to newly created frame
+* \param[in] original    A reference frame that can be used to fill in auxilary information like format, width, height, bpp, stride (if applicable)
+* \param[in] new_bpp     New value for bits per pixel for the allocated frame
```
bpp and stride could be omitted here.
Actually the reason we have w,h is to conform with video frames

ev-mp commented on include/librealsense2/h/rs_sensor.h:
```diff
@@ -79,6 +79,9 @@ typedef enum rs2_format
     RS2_FORMAT_DISTANCE        , /**< 32-bit float-point depth distance value.  */
     RS2_FORMAT_MJPEG           , /**< Bitstream encoding for video in which an image of each frame is encoded as JPEG-DIB   */
+    RS2_FORMAT_Y8I             , /**< 8-bit per pixel interleaved. 8-bit left, 8-bit right.  */
+    RS2_FORMAT_INZI            , /**< 32-bit infra-red image followed by depth data.  */
```
INZI stands for multi-planar Depth 16bit + IR 10bit
https://linuxtv.org/downloads/v4l-dvb-apis/uapi/v4l/pixfmt-inzi.html

INVI - is IR stream only. Don't remember if 8 or 10bit

arilowen replied:
Thanks

ev-mp commented on include/librealsense2/h/rs_sensor.h:
```diff
@@ -79,6 +79,9 @@ typedef enum rs2_format
     RS2_FORMAT_Y10BPACK        , /**< 16-bit per-pixel grayscale image unpacked from 10 bits per pixel packed ([8:8:8:8:2222]) grey-scale image. The data is unpacked to LSB and padded with 6 zero bits */
     RS2_FORMAT_DISTANCE        , /**< 32-bit float-point depth distance value.  */
     RS2_FORMAT_MJPEG           , /**< Bitstream encoding for video in which an image of each frame is encoded as JPEG-DIB   */
+    RS2_FORMAT_Y8I             , /**< 8-bit per pixel interleaved. 8-bit left, 8-bit right.  */
```
Why do you need to expose this to public API ? 
This format is an implementation detail and should remain internal. From the user's perspective he will ask for left &| right IR, and the coce will decide which 4cc pass to kernel

ev-mp commented on include/librealsense2/hpp/rs_processing.hpp:
```diff
@@ -45,6 +45,33 @@ namespace rs2
+        */
+        frame allocate_motion_frame(const stream_profile& profile,
+            const frame& original,
+            int new_bpp = 0,
```
bpp/stride as above

ev-mp commented on include/librealsense2/hpp/rs_processing.hpp:
```diff
@@ -45,6 +45,33 @@ namespace rs2
+        * \param[in] new_bpp     Frame bit per pixel to create.
+        * \param[in] new_width   Frame width to create.
+        * \param[in] new_height  Frame height to create.
+        * \param[in] new_stride  Frame stride to crate.
```
cr**e**ate

ev-mp commented on src/core/advanced_mode.h:
```diff
@@ -9,6 +9,7 @@
 #define RS400_ADVANCED_MODE_HPP
 #include "ds5/advanced_mode/presets.h"
 #include "../../include/librealsense2/h/rs_advanced_mode_command.h"
+#include "ds5/ds5-device.h"
```
Is this include necessary ?

ev-mp commented on src/ds5/ds5-color.cpp:
```diff
@@ -19,6 +19,10 @@
 
+#include "proc/color-formats-converter.h"
+#include "proc/identity-processing-block.h"
+#include "functional"
```
use <...> for STL headers

ev-mp commented on src/ds5/ds5-color.cpp:
```diff
@@ -41,67 +45,59 @@ namespace librealsense
-            std::unique_ptr<frame_timestamp_reader>(new global_timestamp_reader(std::move(ds5_timestamp_reader_metadata), _tf_keeper, enable_global_time_option)));
-
-        _color_device_idx = add_sensor(color_ep);
+        auto color_ep = std::make_shared<uvc_sensor>("Raw RGB Sensor", backend.create_uvc_device(color_devices_info.front()),
```
There are several occurrences in SDK for comparing strings against "RGB Sensor".
Pls run a search to verify no regressions

ev-mp commented on src/ds5/ds5-color.cpp:
```diff
@@ -154,11 +150,34 @@ namespace librealsense
+        smart_color_ep->register_processing_block({ {RS2_FORMAT_UYVY} }, { {RS2_FORMAT_BGRA8, RS2_STREAM_COLOR} }, []() { return std::make_shared<color_formats_converter>(RS2_FORMAT_UYVY, RS2_FORMAT_BGRA8); });
+        smart_color_ep->register_processing_block({ {RS2_FORMAT_UYVY} }, { {RS2_FORMAT_UYVY, RS2_STREAM_COLOR} }, []() { return std::make_shared<identity_processing_block>(); });
+
+        smart_color_ep->register_processing_block({ {RS2_FORMAT_RAW16} }, { {RS2_FORMAT_RAW16, RS2_STREAM_COLOR} }, []() { return std::make_shared<color_formats_converter>(RS2_FORMAT_RAW16, RS2_FORMAT_RAW16); });
```
Isn't Raw to Raw == identity ?
In case there is some subtle transform, like changing endianness - pls put a comment

aangerma commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+        for (auto&& req : requests)
+        {
+            // for motion profile, request is satisfied if it has the same format and stream type.
+            if (Is<motion_stream_profile, stream_profile_interface>(req))
```
why do you need special treatment to motion?

arilowen replied:
Motion profiles are differentiated by stream type, for instance RS2_STREAM_ACCEL, while their format is identical.
Regular video / depth profiles are differentiated by their formats.
In case of zero order, two profiles, Y8 (IR) and Z16(depth) are compared to a single profile, Z16(depth). Which will lead to comparing IR and depth which will return false, meaning this factory does not satisfies the request.

ev-mp commented on src/ds5/ds5-motion.cpp:
```diff
@@ -178,75 +183,138 @@ namespace librealsense
+
+        // register pre-processing
+        smart_hid_ep->register_processing_block(
+            { {RS2_FORMAT_MOTION_XYZ32F, RS2_STREAM_ACCEL} },
```
Is the motion_correction modules implemented here?

arilowen replied:
Please see the lambda in line 210 in the same file.

ev-mp commented on src/ds5/ds5-timestamp.cpp:
```diff
@@ -64,25 +71,28 @@ namespace librealsense
             pin_index = 1;
 
-        if(_has_metadata[pin_index] && fo.metadata_size > platform::uvc_header_size)
+        if(_has_metadata[pin_index] && f->additional_data.metadata_size > platform::uvc_header_size)
```
Add check for (f==nullptr) - dynamic_ptr_cast

ev-mp commented on src/ds5/ds5-timestamp.cpp:
```diff
@@ -157,30 +167,31 @@ namespace librealsense
         // The timewstamp shall be trimmed back to 32 bit to allow HID/UVC intra-stream sync
         // See ds5_iio_hid_timestamp_reader description
-        auto timestamp = *((uint32_t*)((const uint8_t*)fo.pixels + timestamp_offset));
+        auto timestamp = *((uint32_t*)((const uint8_t*)f->get_frame_data() + timestamp_offset));
```
Check for f==nullptr

ev-mp commented on src/image.cpp:
```diff
@@ -1033,6 +1009,74 @@ namespace librealsense
+        case RS2_FORMAT_BGRA8:
+            unpack_yuy2<RS2_FORMAT_BGRA8>(d, s, w, h, actual_size);
+            break;
+        }
```
Add default case to handle exceptions

ev-mp commented on src/image.cpp:
```diff
@@ -1033,6 +1009,74 @@ namespace librealsense
+            break;
+        case RS2_FORMAT_BGRA8:
+            unpack_uyvy<RS2_FORMAT_BGRA8>(d, s, w, h, actual_size);
+            break;
```
Default:

ev-mp commented on src/image.cpp:
```diff
@@ -1033,6 +1009,74 @@ namespace librealsense
+        unpack_gyro_axes<RS2_FORMAT_MOTION_XYZ32F>(dest, source, width, height, actual_size);
+    }
+
+    void align_l500_y8_optimized(byte * const dest[], const byte * source, int width, int height, int actual_size)
```
PLs consider to rename `align_l500..` to `unpack_..` for consistency.
Also replace  the model name in the signature with a meaningful action

ev-mp commented on src/ivcam/sr300.cpp:
```diff
@@ -339,39 +339,41 @@ namespace librealsense
             {
                 LOG_WARNING("UVC metadata payloads are not available for stream "
-                    << std::hex << mode.pf->fourcc << std::dec << (mode.profile.format)
+                    //<< std::hex << mode.pf->fourcc << std::dec << (mode.profile.format)
```
It is very useful for user to have the fourcc notification for debugging

ev-mp commented on src/ivcam/sr300.cpp:
```diff
@@ -339,39 +339,41 @@ namespace librealsense
         std::lock_guard<std::recursive_mutex> lock(_mtx);
 
-        if (has_metadata_fc(fo))
+        auto f = std::dynamic_pointer_cast<librealsense::frame>(frame);
```
Move the dynami_cast under "if" statement and check for nullptr

ev-mp commented on src/ivcam/sr300.h:
```diff
@@ -50,11 +50,13 @@ namespace librealsense
             // Timestamps are encoded within the first 32 bits of the image and provided in 10nsec units
-            uint32_t rolling_timestamp = *reinterpret_cast<const uint32_t *>(fo.pixels);
+            auto f = std::dynamic_pointer_cast<librealsense::frame>(frame);
+            uint32_t rolling_timestamp = *reinterpret_cast<const uint32_t *>(f->get_frame_data());
```
f!=nullptr

ev-mp commented on src/ivcam/sr300.h:
```diff
@@ -69,15 +71,17 @@ namespace librealsense
+        rs2_timestamp_domain get_frame_timestamp_domain(std::shared_ptr<frame_interface> frame) const override
         {
-            if(fo.metadata_size >= platform::uvc_header_size )
+            auto f = std::dynamic_pointer_cast<librealsense::frame>(frame);
```
Same

ev-mp commented on src/ivcam/sr300.h:
```diff
@@ -93,21 +97,23 @@ namespace librealsense
             // Metadata support for a specific stream is immutable
             const bool has_md_ts = [&]{ std::lock_guard<std::recursive_mutex> lock(_mtx);
-                return ((fo.metadata != nullptr) && (fo.metadata_size >= platform::uvc_header_size) && ((byte*)fo.metadata)[0] >= platform::uvc_header_size);
+                return ((f->additional_data.metadata_blob.data() != nullptr) && (f->additional_data.metadata_size >= platform::uvc_header_size) && ((byte*)f->additional_data.metadata_blob.data())[0] >= platform::uvc_header_size);
```
same as before

ev-mp commented on src/ds5/ds5-color.cpp:
```diff
@@ -41,67 +45,59 @@ namespace librealsense
-        color_ep->register_pu(RS2_OPTION_GAIN);
-        color_ep->register_pu(RS2_OPTION_GAMMA);
-        color_ep->register_pu(RS2_OPTION_SHARPNESS);
+        smart_color_ep->register_pu(RS2_OPTION_BRIGHTNESS);
```
Pls consider to reuse the regular `color_ep` name to designate the new aggregated sensor. Renaming the"old" sensor as "raw_sensor" could be sufficient

alowenst01 committed:
wip - fix github comments

ev-mp commented on src/l500/l500-depth.h:
```diff
@@ -161,9 +159,11 @@ namespace librealsense
+            auto results = synthetic_sensor::init_stream_profiles();
             for (auto p : results)
             {
+                auto sp = As<video_stream_profile, stream_profile_interface>(p);
```
Is this var used ?

ev-mp commented on src/proc/CMakeLists.txt:
```diff
@@ -40,9 +44,13 @@ target_sources(${LRS_TARGET}
         "${CMAKE_CURRENT_LIST_DIR}/rates-printer.h"
         "${CMAKE_CURRENT_LIST_DIR}/zero-order.h"
         "${CMAKE_CURRENT_LIST_DIR}/units-transform.h"
+        "${CMAKE_CURRENT_LIST_DIR}/l500-transform.h"
```
The processing block names should remain generic -  pls replace the SKU with functionality description

ev-mp commented on src/proc/color-formats-converter.cpp:
```diff
@@ -0,0 +1,49 @@
+    {
+        _source_stream_profile = p;
+        _target_stream_profile = p.clone(p.stream_type(), p.stream_index(), _target_format);
+        _traget_bpp = get_image_bpp(_target_format) / 8;
```
typo _target_pp

ev-mp commented on src/proc/color-formats-converter.cpp:
```diff
@@ -0,0 +1,49 @@
+    case RS2_FORMAT_MJPEG:
+        unpack_mjpeg(planes, (const byte*)f.get_data(), vf.get_width(), vf.get_height(), vf.get_height() * vf.get_width() * _traget_bpp);
+        break;
+    }
```
default:

ev-mp commented on src/proc/l500-transform.cpp:
```diff
@@ -0,0 +1,77 @@
+            target_spi->set_unique_id(p.unique_id());
+        }
+
+        auto vf = f.as<rs2::video_frame>();
```
Verify vf as motion_frame could potentially arrive here

ev-mp commented on src/proc/l500-transform.h:
```diff
@@ -0,0 +1,22 @@
+        rs2_extension _extension_type;
+        int _traget_bpp = 0;
+    };
+}
```
Add Enter

ev-mp commented on src/proc/motion-transform.cpp:
```diff
@@ -0,0 +1,57 @@
+    motion_transform::motion_transform(rs2_format target_format, rs2_stream target_stream, frame_callback_ptr cb)
+        : motion_transform("Motion Transform")
+    {
+        _callback = cb;
```
Can be moved into the  initialization list

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+    {
+
+        auto&& rhs_src_fmts = rhs.get_source_info();
+        for (auto src : _source_info)
```
Use const ref

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+        }
+
+        auto&& rhs_tgt_fmts = rhs.get_target_info();
+        for (auto tgt : _target_info)
```
Same as in L39

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+        _source_info(from), _target_info(to), generate_processing_block(generate_func)
+    {}
+
+    bool processing_block_factory::operator==(const processing_block_factory & rhs)
```
make const

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+        return true;
+    }
+
+    bool processing_block_factory::has_source(std::shared_ptr<stream_profile_interface> source)
```
make const

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+
+    bool processing_block_factory::has_source(std::shared_ptr<stream_profile_interface> source)
+    {
+        for (auto s : _source_info)
```
Same as in L39

arilowen replied:
Unable to binding lvalue objects to rvalue references.

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+            {
+                if (std::find_if(begin(tgts), end(tgts), [&req](auto tgt) {
+                    return req->get_format() == tgt.format &&
+                        req->get_stream_type() == tgt.stream;
```
What about FPS? Is it verified elsewhere

arilowen replied:
The processing block factory targets are initialized with generic stream profiles which hold only formats and stream type.
They do not hold any resolution nor fps data.
The purpose of stream_profile struct in processing_block_factory is only for consistency and readability.
The fps and all the rest of the requests info are mapped from the request to the relevant source profile via _target_to_source_profiles_map in resolve_request.

ev-mp commented on src/proc/processing-blocks-factory.cpp:
```diff
@@ -25,4 +27,71 @@ namespace librealsense
+        return false;
+    }
+
+    stream_profiles processing_block_factory::find_satisfied_requests(stream_profiles requests)
```
make const

ev-mp commented on src/proc/synthetic-stream.cpp:
```diff
@@ -410,4 +430,69 @@ namespace librealsense
+
+    processing_block & composite_processing_block::get(rs2_option option)
+    {
+        // Find the first block which supports the option.
```
This is the one we discussed - when the composite PB includes more than one module with a given option, (decimation/temporal filter, e.g.) so when the option applied it should be applied exclusively to the designated filter.
The same applies to get : the CPB should expose options per internal PB

ev-mp commented on src/proc/synthetic-stream.cpp:
```diff
@@ -410,4 +430,69 @@ namespace librealsense
+    {
+        // Each processing block will process the preceding processing block output frame.
+        size_t i = 0;
+        for (i = 1; i < _processing_blocks.size(); i++)
```
You can initialize i in the `for` statement

arilowen replied:
It is on purpose, the "i" variable is addressed after the loop.

ev-mp commented on src/proc/synthetic-stream.h:
```diff
@@ -144,6 +152,56 @@ namespace librealsense
+                // will only read from the currently selected
+                // block, setting an option will propogate
+                // to all blocks in the group
+                for (int i = 0; i < _parent->_processing_blocks.size(); i++)
```
See the point raised regarding CPB dispatching the get/set to the appropriate internal PB above

ev-mp commented on src/sensor.cpp:
```diff
@@ -127,141 +141,23 @@ namespace librealsense
-        {
-            if (_is_streaming)
-                uvc_sensor::stop();
+    //rs2_format sensor_base::advanced_to_backend_format(rs2_format format) const
```
Remove if not needed anymore

ev-mp commented on src/sensor.cpp:
```diff
@@ -430,141 +288,90 @@ namespace librealsense
+                    frame_holder fh = _source.alloc_frame(stream_to_frame_types(req_profile_base->get_stream_type()), width * height * bpp / 8, fr->additional_data, requires_processing);
+                    if (fh.frame)
+                    {
+                        memcpy((void*)fh->get_frame_data(), fr->data.data(), sizeof(byte)*fr->data.size());
```
Should be checked with D465/JPEG format that is zero-padded.

ev-mp commented on src/sensor.cpp:
```diff
@@ -685,6 +517,43 @@ namespace librealsense
+            auto profile = std::make_shared<video_stream_profile>(p);
+            profile->set_dims(p.width, p.height);
+            profile->set_stream_type(fourcc_to_rs2_stream(p.format));
+            profile->set_stream_index(0);
```
D400 infrared use indexes 1,2

ev-mp commented on src/sensor.cpp:
```diff
@@ -1128,20 +946,22 @@ namespace librealsense
         std::lock_guard<std::recursive_mutex> lock(_mtx);
 
-        if (has_metadata(mode, fo.metadata, fo.metadata_size))
+        auto f = std::dynamic_pointer_cast<librealsense::frame>(frame);
```
Move f under "if" statement

ev-mp reviewed (commented):
I went through till synthetic sensor

alowenst01 committed:
wip - fix github comments

dorodnic committed:
Bump version

dorodnic committed:
Bump firmware version to 5.11.15

dorodnic committed:
Fixing tare calibration UI

ev-mp commented on src/proc/motion-transform.cpp:
```diff
@@ -0,0 +1,56 @@
+        planes[0] = (byte*)ret.get_data();
+
+        if (_target_stream == RS2_STREAM_ACCEL)
+            unpack_acceleration_axes(planes, (const byte*)f.get_data(), width, height, width * height * _traget_bpp);
```
Those two should be quite similar - both produce 3 float, so maybe those could be merged

ev-mp commented on src/proc/motion-transform.h:
```diff
@@ -0,0 +1,29 @@
+        rs2::stream_profile _target_stream_profile;
+        rs2::stream_profile _source_stream_profile;
+
+        int _traget_bpp = 1;
```
typo

ev-mp commented on src/proc/rotation-transform.cpp:
```diff
@@ -0,0 +1,86 @@
+            // Set the unique ID as the original frame.
+            // The frames are piped through a syncer and must have the origin UID.
+            auto target_spi = (stream_profile_interface*)_target_stream_profile.get()->profile;
+            target_spi->set_unique_id(p.unique_id());
```
Is it possible to pass this condition as a flag to `_target_stream_profile = p.clone(p.s` ? so it will be performed during clone ?

dorodnic committed:
Merge pull request #4909 from aangerma/development

Added to align sse support on distortion parameters and fixed bug on playback

ev-mp commented on src/sensor.h:
```diff
@@ -266,6 +330,29 @@ namespace librealsense
+        };
+        std::map<uint32_t, rs2_stream> _fourcc_to_rs2_stream = {
+            {rs_fourcc('Y','U','Y','2'), RS2_STREAM_COLOR},
+            {rs_fourcc('U','Y','V','Y'), RS2_STREAM_COLOR},
```
UYUV should represent the synthetic color for IR stream

ev-mp commented on src/types.h:
```diff
@@ -641,6 +657,14 @@ namespace librealsense
             (a.stream == b.stream);
     }
 
+    inline bool operator<(const stream_profile & lhs,
```
What does this mean `lhs.format < rhs.format` ? How and where is it applied ?

dorian3d committed:
rs_sensor: documentation of relocalization functions updated

dorian3d committed:
rs_device: doxygen of tm2 has links to sensors

dorodnic committed:
Merge pull request #4927 from dorian3d/reloc-doxygen

Update doxygen of pose_sensor

dorian3d committed:
rs_types: new notification category for pose relocalizations

dorodnic committed:
Merge pull request #4825 from matkatz/l500-firmware-update

Firmware update support

dorodnic committed:
Merge pull request #4925 from dorodnic/tare_fix

Fixing Tare Calibration UI + Bump version

dorian3d committed:
tm-device: relocalization raises relocalization event

alowenst01 committed:
wip - fix github comments

alowenst01 committed:
wip - fix processing block factory satisfaction condition

alowenst01 committed:
wip - fix sergey's comments

Commit:
Merge branch 'master' into dev

Change-Id: I8e2edbc2d0f5aea1be4592756c0f3339df1215d3

alowenst01 committed:
wip - fix sergey's comments2

aangerma committed:
Merge pull request #4941 from IntelRealSense/development

Merge of 2.29 to master

dorian3d committed:
csharp: relocalization notification category

alowenst01 committed:
wip - fix sergey's comments3

radfordi committed:
rs-ar-basic: fix extrinsic pose to camera transformation

Found by user @JBBee: https://github.com/IntelRealSense/librealsense/issues/4883#issuecomment-535597865

cabelo committed:
illustration for project

cabelo committed:
illustration for project

cabelo committed:
illustration for project

radfordi committed:
ar-basic: Fix the extrinsics in the readme to match the fixed code

dorodnic committed:
Merge pull request #4945 from radfordi/ar-basic-extrinsics

rs-ar-basic: fix extrinsic pose to camera transformation

dorodnic committed:
Merge pull request #4936 from dorian3d/reloc-notification

Notification category for Relocalization

alowenst01 committed:
wip - fix sergey's comments 3

alowenst01 committed:
wip - refactor motion transform

alowenst01 committed:
wip - refactor colors and rotation converters

alowenst01 committed:
wip - refactor motion transform

AnnaRomanov committed:
rename foreach to foreachRS where is was missing

AnnaRomanov committed:
changed foreachRS to foreach_rs

alowenst01 committed:
wip - fix recording

alowenst01 committed:
wip - add fisheye support

alowenst01 committed:
wip - add fisheye

Commit:
Added an overload to config_value's conversion operator, allowing correct conversion from config_value to std::string

alowenst01 committed:
wip - add sr300

AnnaRomanov committed:
Changed tabs to spaces

dorodnic committed:
Merge pull request #5028 from AnnaRomanov/username_viewer_bug

Fix default path bug in viewer

alowenst01 committed:
wip - add ds5u support

alowenst01 committed:
wip - expose processing block options to sensor

AnnaRomanov committed:
Free name string to avoid memory leaks in hid sensor

dorodnic committed:
Merge pull request #4987 from AnnaRomanov/foreach_rename

Rename foreach to foreachRS

alowenst01 committed:
wip - add support for platform camera and fix l500 streaming issue

alowenst01 committed:
wip - clean native pixel formats

alowenst01 committed:
wip - clean native pixel format and std::algorithm refactoring

dorodnic committed:
Updating Mac OS documentation

alowenst01 committed:
wip - fix fisheye auto exposure

alowenst01 committed:
wip - refactor references and copy vars

alowenst01 committed:
wip - refactor processing blocks and image.cpp

dorodnic committed:
Merge pull request #4953 from cabelo/img-realsense

illustration for project

dorodnic committed:
Merge pull request #5065 from AnnaRomanov/d435i_win_mem_leak

Fix memory leak in hid sensor

alowenst01 committed:
wip - fix sr300 y16 format

alowenst01 committed:
wip - record after pre-processing fix

alowenst01 committed:
resolve request failure throw an informative message

alowenst01 committed:
merge with development

alowenst01 committed:
synthetic sensor support notification fix
